lib/Dialect/SystemC/SystemCTypes.cpp

```diff
@@ -101,4 +199,113 @@ void SystemCDialect::registerTypes() {
 #define GET_TYPEDEF_LIST
 #include "circt/Dialect/SystemC/SystemCTypes.cpp.inc"
       >();
+  addTypes<UIntType, IntType, BigIntType, BigUIntType, UIntBaseType,
+           IntBaseType, SignedType, UnsignedType, BitVectorBaseType,
+           BitVectorType, LogicVectorBaseType, LogicVectorType>();
+}
+
+//===----------------------------------------------------------------------===//
+// Custom type parsers and printers
+//===----------------------------------------------------------------------===//
+
+template <typename Ty>
+static LogicalResult parseIntegerOrBitVector(DialectAsmParser &parser,
+                                             Type &type) {
+  unsigned width;
+  auto *ctxt = parser.getContext();
+
+  if (parser.parseLess() || parser.parseInteger(width) || parser.parseGreater())
+    return failure();
+
+  type = Ty::get(ctxt, width);
+  return success();
+}
+
+static mlir::OptionalParseResult customTypeParser(DialectAsmParser &parser,
+                                                  StringRef mnemonic,
+                                                  llvm::SMLoc loc, Type &type) {
+  auto *ctxt = parser.getContext();
+
+  if (mnemonic == IntBaseType::getMnemonic()) {
+    type = IntBaseType::get(ctxt);
+    return success();
+  }
```

Comment:
An LLVM `StringSwitch` _might_ clean this up. It's kind of a weird API, though.

Reply:
`StringSwitch` eagerly executes both arguments of its `Case` calls of all cases up to the matching one, i.e., you cannot do anything that has side effects in these calls (e.g. parsing). There does not seem to be a variant that takes lambdas and lazily executes them, which would be really nice. Even if I split it up to two `StringSwitch` statements, one for the types that don't have a bitwidth and the ones that have one and do the parsing outside of the `StringSwitch`, we end up unnecessarily `get`ting these types (and the code doesn't really look cleaned up anymore as well IMO).